File: app/src/main/java/com/github/teranes10/androidutils/utils/location/LocationManagerProvider.java
package com.github.teranes10.androidutils.utils.location;

import android.Manifest;
import android.content.Context;
import android.content.pm.PackageManager;
import android.location.Criteria;
import android.location.LocationListener;
import android.location.LocationManager;
import android.util.Log;

import androidx.core.app.ActivityCompat;

public class LocationManagerProvider extends LocationProvider {
    private static final String TAG = "LocationManagerProvider";
    private final Context _ctx;
    private final LocationManager mLocationManager;
    private final LocationListener mLocationListener;

    public LocationManagerProvider(Context ctx, ILocationListener listener) {
        Log.i(TAG, "LocationManagerProvider: ");
        this._ctx = ctx;
        mLocationManager = (LocationManager) _ctx.getApplicationContext().getSystemService(Context.LOCATION_SERVICE);
        mLocationListener = location -> {
            if (listener != null) {
                listener.onLocationChanged(location);
            }
        };
    }

    @Override
    public void startUpdates(int intervalMillis, int minIntervalMills, int minDistance) {
        if (mLocationManager == null) {
            return;
        }

        final boolean gpsEnabled = mLocationManager.isProviderEnabled(LocationManager.GPS_PROVIDER);
        if (!gpsEnabled) {
            return;
        }

        if (ActivityCompat.checkSelfPermission(_ctx, Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED ||
                ActivityCompat.checkSelfPermission(_ctx, Manifest.permission.ACCESS_COARSE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
            return;
        }

        Criteria criteria = new Criteria();
        criteria.setAccuracy(Criteria.ACCURACY_FINE);
        mLocationManager.getBestProvider(criteria, true);
        mLocationManager.requestLocationUpdates(LocationManager.GPS_PROVIDER, minIntervalMills, minDistance, mLocationListener);
        mLocationManager.requestLocationUpdates(LocationManager.NETWORK_PROVIDER, minIntervalMills, minDistance, mLocationListener);
    }

    @Override
    public void stopUpdates() {
        if (mLocationManager == null) {
            return;
        }

        mLocationManager.removeUpdates(mLocationListener);
    }
}

File: app/src/main/java/com/github/teranes10/androidutils/utils/location/LocationManagerProvider.kt
package com.github.teranes10.androidutils.utils.location

import android.annotation.SuppressLint
import android.content.Context
import android.hardware.Sensor
import android.hardware.SensorManager
import android.location.LocationListener
import com.github.teranes10.androidutils.models.Outcome

class LocationManagerProvider(
    context: Context, listener: ILocationListener?, sensorType: Int = Sensor.TYPE_LINEAR_ACCELERATION,
    sensorDelay: Int = SensorManager.SENSOR_DELAY_GAME,
    lowPassFilterAlpha: Float = 0.2f,
    magnitudeHistorySize: Int = 50
) : LocationProvider(
    context,
    sensorType = sensorType,
    sensorDelay = sensorDelay,
    lowPassFilterAlpha = lowPassFilterAlpha,
    magnitudeHistorySize = magnitudeHistorySize
) {
    private val locationListener = LocationListener { location -> listener?.onLocationChanged(location) }

    @SuppressLint("MissingPermission")
    override fun startUpdates(intervalMillis: Long, minIntervalMillis: Long, minDistance: Float): Outcome<*> {
        val result = super.startUpdates(intervalMillis, minIntervalMillis, minDistance)
        if (result.failure) {
            return result
        }

        for (provider in activeProviders) {
            locationManager.requestLocationUpdates(provider, minIntervalMillis, minDistance, locationListener, handler!!.looper)
        }

        return result
    }

    override fun stopUpdates() {
        locationManager.removeUpdates(locationListener)
        super.stopUpdates()
    }
}
